api/base/base_controller.go

```diff
@@ -0,0 +1,336 @@
+			Handler: c.ListObjects,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodDelete,
+				Path:   c.resourceBaseURL,
+			},
+			Handler: c.DeleteObjects,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodDelete,
+				Path:   fmt.Sprintf("%s/{%s}", c.resourceBaseURL, PathParamID),
+			},
+			Handler: c.DeleteSingleObject,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodPatch,
+				Path:   fmt.Sprintf("%s/{%s}", c.resourceBaseURL, PathParamID),
+			},
+			Handler: c.PatchObject,
+		},
+	}
+}
+
+func (c *Controller) CreateObject(r *web.Request) (*web.Response, error) {
+	ctx := r.Context()
+	log.C(ctx).Debugf("Creating new %s", c.objectType)
+
+	var createInterceptor extension.CreateInterceptor
+	if len(c.CreateInterceptorProviders) > 0 {
+		createInterceptor = extension.UnionCreateInterceptor(c.CreateInterceptorProviders)()
+	}
+
+	onTransaction := func(ctx context.Context, txStorage storage.Warehouse, newObject types.Object) error {
+		id, err := txStorage.Create(ctx, newObject)
+		if err != nil {
+			return util.HandleStorageError(err, string(c.objectType))
+		}
+		newObject.SetID(id)
+		return nil
+	}
+	if createInterceptor != nil {
+		onTransaction = createInterceptor.OnTxCreate(onTransaction)
+	}
+
+	onAPI := func(ctx context.Context, obj types.Object) (types.Object, error) {
+		if err := c.repository.InTransaction(ctx, func(ctx context.Context, txStorage storage.Warehouse) error {
+			return onTransaction(ctx, txStorage, obj)
+		}); err != nil {
+			return nil, err
+		}
+		return obj, nil
+	}
+	if createInterceptor != nil {
+		onAPI = createInterceptor.OnAPICreate(onAPI)
+	}
+	result := c.objectBlueprint()
+	if err := util.BytesToObject(r.Body, result); err != nil {
+		return nil, err
+	}
+
+	if result.GetID() == "" {
+		UUID, err := uuid.NewV4()
+		if err != nil {
+			return nil, fmt.Errorf("could not generate GUID for %s: %s", c.objectType, err)
+		}
+		result.SetID(UUID.String())
+	}
+	currentTime := time.Now().UTC()
+	result.SetCreatedAt(currentTime)
+	result.SetUpdatedAt(currentTime)
+
+	result, err := onAPI(ctx, result)
+	if err != nil {
+		return nil, err
+	}
+	return web.NewJSONResponse(http.StatusCreated, result)
+}
+
+func (c *Controller) DeleteObjects(r *web.Request) (*web.Response, error) {
+	ctx := r.Context()
+	log.C(ctx).Debugf("Deleting %ss...", c.objectType)
+	var deleteInterceptor extension.DeleteInterceptor
+	if len(c.DeleteInterceptorProviders) > 0 {
+		deleteInterceptor = extension.UnionDeleteInterceptor(c.DeleteInterceptorProviders)()
+	}
+
+	transactionOperation := func(ctx context.Context, txStorage storage.Warehouse, deletionCriteria ...query.Criterion) (types.ObjectList, error) {
+		return c.repository.Delete(ctx, c.objectType, deletionCriteria...)
+	}
+	if deleteInterceptor != nil {
+		transactionOperation = deleteInterceptor.OnTxDelete(transactionOperation)
+	}
+
+	apiOperation := func(ctx context.Context, deletionCriteria ...query.Criterion) (types.ObjectList, error) {
```

Comment:
this is pretty hard to read/follow - I would propose to extract the code here in something that implements an interfaces which is also implemented by the `type InterceptDeleteOnAPIFunc func(ctx context.Context, deletionCriteria ...query.Criterion) (types.ObjectList, error)` and `type InterceptDeleteOnTxFunc func(ctx context.Context, txStorage storage.Warehouse, deletionCriteria ...query.Criterion) (types.ObjectList, error)
` funcs. Here are code snippets:

```
package extension

import (
	"context"
	"fmt"

	"github.com/Peripli/service-manager/pkg/types"

	"github.com/Peripli/service-manager/pkg/query"

	"github.com/Peripli/service-manager/storage"
)

type namedDeleteAPIFunc struct {
	Name string
	Func func(InterceptDeleteOnAPIFunc) InterceptDeleteOnAPIFunc
}

type namedDeleteTxFunc struct {
	Name string
	Func func(InterceptDeleteOnTxFunc) InterceptDeleteOnTxFunc
}

//TODO this has ONAPI in the name and has an array of TX Funcs that are structs. Naming / abstraction needs improvements
type deleteInterceptorChain struct {
	DeleteHookOnAPIFuncs []*namedDeleteAPIFunc
	DeleteHookOnTxFuncs  []*namedDeleteTxFunc
}

func (c *deleteInterceptorChain) OnAPIDelete(f InterceptDeleteOnAPIFunc) InterceptDeleteOnAPIFunc {
	for i := range c.DeleteHookOnAPIFuncs {
		f = c.DeleteHookOnAPIFuncs[len(c.DeleteHookOnAPIFuncs)-1-i].Func(f)
	}
	return f
}

func (c *deleteInterceptorChain) OnTxDelete(f InterceptDeleteOnTxFunc) InterceptDeleteOnTxFunc {
	for i := range c.DeleteHookOnTxFuncs {
		f = c.DeleteHookOnTxFuncs[len(c.DeleteHookOnTxFuncs)-1-i].Func(f)
	}
	return f
}

// NewDeleteInterceptorChain returns a function which chains all delete interceptors, sorts them and wraps them into one.
//TODO this doesnt need to return a func
func NewDeleteInterceptorChain(providers []DeleteInterceptorProvider) *deleteInterceptorChain {
	c := &deleteInterceptorChain{}
	c.DeleteHookOnAPIFuncs = make([]*namedDeleteAPIFunc, 0, len(providers))
	c.DeleteHookOnTxFuncs = make([]*namedDeleteTxFunc, 0, len(providers))

	for _, p := range providers {
		hook := p.Provide()
		positionAPIType := PositionNone
		positionTxType := PositionNone
		nameAPI := ""
		nameTx := ""

		if orderedProvider, isOrdered := p.(Ordered); isOrdered {
			positionAPIType, nameAPI = orderedProvider.PositionAPI()
			positionTxType, nameTx = orderedProvider.PositionTx()
		}

		c.insertAPIFunc(positionAPIType, nameAPI, &namedDeleteAPIFunc{
			Name: p.Name(),
			Func: hook.OnAPIDelete,
		})
		c.insertTxFunc(positionTxType, nameTx, &namedDeleteTxFunc{
			Name: p.Name(),
			Func: hook.OnTxDelete,
		})
	}

	return c
}

// DeleteInterceptorProvider provides DeleteInterceptorChain for each request
//go:generate counterfeiter . DeleteInterceptorProvider
type DeleteInterceptorProvider interface {
	// TODO name should be on DeleteInterceptor and not on the provider, provide first, order by name after
	Named
	Provide() DeleteInterceptor
}

//TODO this needs a better name
type InterceptDeleteOnAPI interface {
	InterceptDeleteOnAPI(context.Context, ...query.Criterion) (types.ObjectList, error)
}

//TODO this needs a better name

// InterceptDeleteOnAPIFunc hook for entity deletion outside of transaction
type InterceptDeleteOnAPIFunc func(ctx context.Context, deletionCriteria ...query.Criterion) (types.ObjectList, error)

func (idf InterceptDeleteOnAPIFunc) InterceptDeleteOnAPI(ctx context.Context, criteria ...query.Criterion) (types.ObjectList, error) {
	return idf(ctx, criteria...)
}

type InterceptDeleteOnTx interface {
	InterceptDeleteOnTx(context.Context, storage.Warehouse, ...query.Criterion) (types.ObjectList, error)
}

// InterceptDeleteOnTxFunc hook for entity deletion in transaction
type InterceptDeleteOnTxFunc func(ctx context.Context, txStorage storage.Warehouse, deletionCriteria ...query.Criterion) (types.ObjectList, error)

// DeleteInterceptor provides hooks on entity deletion
//go:generate counterfeiter . DeleteInterceptor
type DeleteInterceptor interface {
	Named
	OnAPIDelete(h InterceptDeleteOnAPIFunc) InterceptDeleteOnAPIFunc
	OnTxDelete(f InterceptDeleteOnTxFunc) InterceptDeleteOnTxFunc
}

//TODO fix reciever names
func (c *deleteInterceptorChain) insertAPIFunc(positionType PositionType, name string, h *namedDeleteAPIFunc) {
	if positionType == PositionNone {
		c.DeleteHookOnAPIFuncs = append(c.DeleteHookOnAPIFuncs, h)
		return
	}
	pos := c.findAPIFuncPosition(c.DeleteHookOnAPIFuncs, name)
	if pos == -1 {
		panic(fmt.Errorf("could not find delete API hook with name %s", name))
	}
	c.DeleteHookOnAPIFuncs = append(c.DeleteHookOnAPIFuncs, nil)
	if positionType == PositionAfter {
		pos = pos + 1
	}
	copy(c.DeleteHookOnAPIFuncs[pos+1:], c.DeleteHookOnAPIFuncs[pos:])
	c.DeleteHookOnAPIFuncs[pos] = h
}

func (c *deleteInterceptorChain) insertTxFunc(positionType PositionType, name string, h *namedDeleteTxFunc) {
	if positionType == PositionNone {
		c.DeleteHookOnTxFuncs = append(c.DeleteHookOnTxFuncs, h)
		return
	}
	pos := c.findTxFuncPosition(c.DeleteHookOnTxFuncs, name)
	if pos == -1 {
		panic(fmt.Errorf("could not find delete transaction hook with name %s", name))
	}
	c.DeleteHookOnTxFuncs = append(c.DeleteHookOnTxFuncs, nil)
	if positionType == PositionAfter {
		pos = pos + 1
	}
	copy(c.DeleteHookOnTxFuncs[pos+1:], c.DeleteHookOnTxFuncs[pos:])
	c.DeleteHookOnTxFuncs[pos] = h
}

func (c *deleteInterceptorChain) findAPIFuncPosition(funcs []*namedDeleteAPIFunc, name string) int {
	for i, f := range funcs {
		if f.Name == name {
			return i
		}
	}

	return -1
}

func (c *deleteInterceptorChain) findTxFuncPosition(funcs []*namedDeleteTxFunc, name string) int {
	for i, f := range funcs {
		if f.Name == name {
			return i
		}
	}

	return -1
}
```

```
type finalDeleteObjectInterceptor struct {
	repository storage.Repository
	objectType types.ObjectType
	txChain    func(extension.InterceptDeleteOnTxFunc) extension.InterceptDeleteOnTxFunc
}

func (final *finalDeleteObjectInterceptor) InterceptDeleteOnTx(ctx context.Context, storage storage.Warehouse, criteria ...query.Criterion) (types.ObjectList, error) {
	return final.repository.Delete(ctx, final.objectType, criteria...)
}

func (final *finalDeleteObjectInterceptor) InterceptDeleteOnAPI(ctx context.Context, criteria ...query.Criterion) (types.ObjectList, error) {
	var result types.ObjectList
	if err := final.repository.InTransaction(ctx, func(ctx context.Context, storage storage.Warehouse) error {
		var err error
		result, err = final.txChain(final.InterceptDeleteOnTx)(ctx, storage, criteria...)
		if err != nil {
			return util.HandleSelectionError(err, string(final.objectType))
		}
	}); err != nil {
		return nil, err
	}
	return result, nil
}

func (c *Controller) DeleteObjects(r *web.Request) (*web.Response, error) {
	ctx := r.Context()
	criteria := query.CriteriaForContext(ctx)
	log.C(ctx).Debugf("Deleting %s...", c.objectType)

	interceptorsChain := extension.NewDeleteInterceptorChain(c.DeleteInterceptorProviders)

	finalInterceptor := &finalDeleteObjectInterceptor{
		repository: c.repository,
		objectType: c.objectType,
		txChain:    interceptorsChain.OnTxDelete,
	}
	// TODO the chain itself is also a DeleteInterceptor - not sure if we can somehow leverage this. Ideas?
	if _, err := interceptorsChain.OnAPIDelete(finalInterceptor.InterceptDeleteOnAPI)(ctx, criteria...); err != nil {
		return nil, err
	}

	return web.NewJSONResponse(http.StatusOK, map[string]string{})
}
```

Reply:
This does not even compile.

Reply:
i just tried it and it compiled just fine when the comment was posted - but even if it didnt compile -  it's a pointer to how to rework the code and a comment, not sure what your expectations here are

Reply:
I reworked it, but didn't see much value in doing this anyway :). In my opinion it is not more readable now.